finality-aleph/src/sync/handler/mod.rs

```diff
@@ -82,6 +83,90 @@ pub trait HandlerTypes {
     type Error;
 }
 
+// This is only required because we don't control block imports and thus we can get notifications about blocks being imported that don't fit in the forest. This struct lets us work around this by manually syncing the forest after such an event.
+//TODO(A0-2984): remove this after legacy sync is excised
+enum MissedImportData {
+    AllGood,
+    MissedImports {
+        highest_missed: BlockNumber,
+        last_sync: BlockNumber,
+    },
+}
+
+impl MissedImportData {
+    pub fn new() -> Self {
+        Self::AllGood
+    }
+
+    pub fn update<B, J, CS>(
+        &mut self,
+        missed: BlockNumber,
+        chain_status: &CS,
+    ) -> Result<(), CS::Error>
+    where
+        B: Block,
+        J: Justification<Header = B::Header>,
+        CS: ChainStatus<B, J>,
+    {
+        use MissedImportData::*;
+        match self {
+            AllGood => {
+                *self = MissedImports {
+                    highest_missed: missed,
+                    last_sync: chain_status.top_finalized()?.header().id().number(),
+                }
+            }
+            MissedImports { highest_missed, .. } => {
+                if *highest_missed < missed {
+                    *highest_missed = missed;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn try_sync<B, I, J, CS>(
+        &mut self,
+        chain_status: &CS,
+        forest: &mut Forest<I, J>,
+    ) -> Result<(), CS::Error>
+    where
+        B: Block,
+        I: PeerId,
+        J: Justification<Header = B::Header>,
+        CS: ChainStatus<B, J>,
+    {
+        use MissedImportData::*;
+        if let MissedImports {
+            highest_missed,
+            last_sync,
+        } = self
+        {
+            let top_finalized = chain_status.top_finalized()?.header().id();
+            // we don't want this to happen too often, but it also cannot be too close to the max forest size, thus semi-random weird looking threshold
+            if *last_sync + 1312 >= top_finalized.number() {
+                return Ok(());
+            }
+            let mut to_import = VecDeque::from(chain_status.children(top_finalized.clone())?);
+            while let Some(header) = to_import.pop_front() {
```

Comment:
This loop will execute pessimistically 1312 + a few times in presence of forks, right? Are we sure this is not too many synchronous operations?